servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ConcurrentUtils.java

```diff
@@ -36,6 +37,49 @@ private ConcurrentUtils() {
         // No instances.
     }
 
+    /**
+     * Acquire a lock that allows reentry and attempts to acquire the lock while it is
+     * held can be detected by {@link #releasePendingReentrantLock(AtomicLongFieldUpdater, long, Object)}.
+     * @param lockUpdater The {@link AtomicLongFieldUpdater} used to control the lock state.
+     * @param owner The owner of the lock object.
+     * @param <T> The type of object that owns the lock.
+     * @return {@code 0} if the acquire was unsuccessful, otherwise an identifier that must be passed to a subsequent
+     * call of {@link #releasePendingReentrantLock(AtomicLongFieldUpdater, long, Object)}.
+     */
+    public static <T> long acquirePendingReentrantLock(final AtomicLongFieldUpdater<T> lockUpdater, final T owner) {
+        final long threadId = Thread.currentThread().getId();
+        for (;;) {
+            final long prevThreadId = lockUpdater.get(owner);
+            if (prevThreadId == 0) {
+                if (lockUpdater.compareAndSet(owner, 0, threadId)) {
+                    return threadId;
+                }
+            } else if (prevThreadId == threadId || prevThreadId == -threadId) {
+                return -threadId;
+            } else if (lockUpdater.compareAndSet(owner, prevThreadId,
+                    prevThreadId > 0 ? -prevThreadId : prevThreadId)) {
+                return 0;
+            }
+        }
+    }
+
+    /**
+     * Release a lock that was previously acquired via
+     * {@link #acquirePendingReentrantLock(AtomicLongFieldUpdater, Object)}.
+     * @param lockUpdater The {@link AtomicLongFieldUpdater} used to control the lock state.
+     * @param acquireId The value returned from the previous call to
+     * {@link #acquirePendingReentrantLock(AtomicLongFieldUpdater, Object)}.
+     * @param owner The owner of the lock object.
+     * @param <T> The type of object that owns the lock.
+     * @return {@code true} if the lock was released, or this method call corresponds to a prior re-entrant call
+     * to {@link #acquirePendingReentrantLock(AtomicLongFieldUpdater, Object)}.
+     */
+    public static <T> boolean releasePendingReentrantLock(final AtomicLongFieldUpdater<T> lockUpdater,
+                                                          final long acquireId, final T owner) {
+        assert acquireId != 0;
+        return acquireId < 0 || lockUpdater.getAndSet(owner, 0) == acquireId;
```

Comment:
nit: consider define a static field for `0` so its a bit less of a magic number

Reply:
done